subsys/net/lib/tftp/tftp_client.c

```diff
@@ -53,6 +53,74 @@ static size_t make_request(uint8_t *buf, int request,
 	return ptr - (char *)buf;
 }
 
+/*
+ * Send Data message to the TFTP Server and receive ACK message from it.
+ */
+static int send_data(int sock, uint32_t block_no, size_t data_size, uint8_t *data_buffer)
+{
+	int ret;
+	int count = 0;
+	struct pollfd fds = {
+		.fd     = sock,
+		.events = ZSOCK_POLLIN,
+	};
+
+	LOG_DBG("Client send data: block no %u, size %u", block_no, data_size);
+
+	/* Send data and poll for first ACK response */
+	sys_put_be16(DATA_OPCODE, data_buffer);
+	sys_put_be16(block_no, data_buffer + 2);
+	do {
+		if (count > TFTP_REQ_RETX) {
+			LOG_ERR("No more retransmits. Exiting");
+			return TFTPC_RETRIES_EXHAUSTED;
+		}
+		ret = send(sock, data_buffer, data_size + TFTP_HEADER_SIZE, 0);
+		if (ret < 0) {
+			LOG_ERR("send() error: %d", -errno);
+			return -errno;
+		}
+		count++;
+	} while (poll(&fds, 1, CONFIG_TFTPC_REQUEST_TIMEOUT) <= 0);
+
+	/* Receive ACK from the TFTP Server. */
+	count = 1;
+	do {
```

Comment:
This is still not great, having to separate loops, impacts the retransmission logic. If for example a duplicate ACK is received, the function will no longer attempt to retransmit the current block.

What I think is needed, is an outer loop, dealing with retransmissions, and an inner loop within, with poll/recv, awaiting for ACKs and processing them. Only if the inner loop times-out, the outer loop should take another iteration, retransmitting the block.

Reply:
That is achievable, but maybe not reusing the buffer, that the sending buffer must be kept intact.